transports/janus_pfunix.c

```diff
@@ -0,0 +1,532 @@
+
+		.is_janus_api_enabled = janus_pfunix_is_janus_api_enabled,
+		.is_admin_api_enabled = janus_pfunix_is_admin_api_enabled,
+
+		.send_message = janus_pfunix_send_message,
+		.session_created = janus_pfunix_session_created,
+		.session_over = janus_pfunix_session_over,
+	);
+
+/* Transport creator */
+janus_transport *create(void) {
+	JANUS_LOG(LOG_VERB, "%s created!\n", JANUS_PFUNIX_NAME);
+	return &janus_pfunix_transport;
+}
+
+
+/* Useful stuff */
+static gint initialized = 0, stopping = 0;
+static janus_transport_callbacks *gateway = NULL;
+
+
+/* Unix Sockets server thread */
+static GThread *pfunix_thread = NULL;
+void *janus_pfunix_thread(void *data);
+
+/* Unix Sockets servers */
+static int pfd = 0, admin_pfd = 0;
+/* Socket pair to notify about the need for outgoing data */
+static int write_fd[2];
+
+/* Unix Sockets client session */
+typedef struct janus_pfunix_client {
+	int fd;					/* Client socket */
+	gboolean admin;			/* Whether this client is for the Admin or Janus API */
+	GAsyncQueue *messages;	/* Queue of outgoing messages to push */
+	gint session_timeout:1;	/* Whether a Janus session timeout occurred in the core */
+} janus_pfunix_client;
+static GHashTable *clients = NULL, *clients_by_fd = NULL;
+static janus_mutex clients_mutex;
+
+
+/* Transport implementation */
+int janus_pfunix_init(janus_transport_callbacks *callback, const char *config_path) {
+	if(g_atomic_int_get(&stopping)) {
+		/* Still stopping from before */
+		return -1;
+	}
+	if(callback == NULL || config_path == NULL) {
+		/* Invalid arguments */
+		return -1;
+	}
+
+	/* This is the callback we'll need to invoke to contact the gateway */
+	gateway = callback;
+
+	/* Read configuration */
+	char filename[255];
+	g_snprintf(filename, 255, "%s/%s.cfg", config_path, JANUS_PFUNIX_PACKAGE);
+	JANUS_LOG(LOG_VERB, "Configuration file: %s\n", filename);
+	janus_config *config = janus_config_parse(filename);
+	if(config != NULL) {
+		/* Handle configuration */
+		janus_config_print(config);
+
+		/* First of all Initialize the socketpair for writeable notifications */
+		if(socketpair(PF_LOCAL, SOCK_STREAM, 0, write_fd) < 0) {
+			JANUS_LOG(LOG_FATAL, "Error creating socket pair for writeable events: %d, %s\n", errno, strerror(errno));
+			return -1;
+		}
+
+		/* Setup the Janus API Unix Sockets server(s) */
+		janus_config_item *item = janus_config_get_item_drilldown(config, "general", "enabled");
+		if(!item || !item->value || !janus_is_true(item->value)) {
+			JANUS_LOG(LOG_WARN, "Unix Sockets server disabled (Janus API)\n");
+		} else {
+			item = janus_config_get_item_drilldown(config, "general", "path");
+			char *pfname = (char *)(item && item->value ? item->value : "/tmp/ux-janusapi");
+			if(strlen(pfname) > 108) {
+				JANUS_LOG(LOG_WARN, "The provided path name (%s) is longer than 108 characters, it will be truncated\n", pfname);
+				pfname[108] = '\0';
+			}
+			/* Create socket */
+			pfd = socket(PF_UNIX, SOCK_SEQPACKET, 0);
```

Comment:
I've been playing around with the idea of a transport using SOCK_DGRAM over AF_UNIX and nearly have some code to show. Some thoughts:
- I don't think listen() is needed. You bind() to an address, and then get the address of the sender when receiving with recvfrom()/recvmsg(). You should be able to use this address to send replies with sendto()/sendmsg()
- SOCK_SEQPACKET over AF_UNIX isn't supported on BSD family (I've been coding on OS X) I don't think.
- sendto()/recvfrom() should be atomic operations as far as the calling code is concerned, i.e. they succeed or fail in one hit. This should reduce the amount of thread synchronization you need for sending or receiving.